perf: optimizing theme config loading (#7913)
shopware/platform: changelog/_unreleased/2025-04-09-optimizing-theme-config-loading.md

* Added `\Shopware\Storefront\Theme\ThemeRuntimeConfigService` to handle theme runtime configurations
* Changed theme configuration loading in the code, used during storefront rendering, to use the new `\Shopware\Storefront\Theme\ThemeRuntimeConfigService`:
  * `\Shopware\Storefront\Theme\ResolvedConfigLoader`
  * `\Shopware\Storefront\Theme\ThemeScripts`
  * `\Shopware\Storefront\Theme\ThemeInheritanceBuilder`
  * `\Shopware\Storefront\Framework\Routing\TemplateDataSubscriber`
* Deprecated `\Shopware\Storefront\Theme\CachedResolvedConfigLoader`, as it is no longer used in the storefront
___
# Upgrade Information
## Theme configuration changes
* Theme configuration used during storefront rendering are now stored in a `theme_runtime_config` table and regenerated on the refresh stage of theme lifecycle.
* The `\Shopware\Storefront\Theme\CachedResolvedConfigLoader` is now deprecated and will be removed in the next major version. Please update the code that directly uses it to use the `\Shopware\Storefront\Theme\ResolvedConfigLoader` instead.
___
# Next Major Version Changes
## Removal of CachedResolvedConfigLoader

> * The `\Shopware\Storefront\Theme\CachedResolvedConfigLoader` will be removed.

Reviewer: ```suggestion
* The `\Shopware\Storefront\Theme\CachedResolvedConfigLoader` was removed.
```

this changelog will be part of 6.8, were this removal already happend

Author: Done

---
title: Optimizing theme config loading
issue: https://github.com/shopware/shopware/issues/7766
---
# Core

> * Changed `\Shopware\Core\System\Snippet\SnippetService` to use `\Shopware\Storefront\Theme\ThemeRuntimeConfigService` to get information about available themes

Reviewer: is this still correct, as you have implemented an event?

Author: fixed

___
# Storefront
* Added new `\Shopware\Storefront\Theme\ThemeRuntimeConfig` entity and `theme_runtime_config` table to store theme runtime configuration
* Added `\Shopware\Storefront\Theme\ThemeRuntimeConfigService` to handle theme runtime configurations

> * Changed `\Shopware\Storefront\Theme\ThemeLifecycleService`, adding optional `$configurationCollection` parameter to the `refreshTheme` method, and deprecating class to be marked as final in the next major version.

Reviewer: ```suggestion
* Changed `\Shopware\Storefront\Theme\ThemeLifecycleService`, adding optional `$configurationCollection` parameter to the `refreshTheme` method
* Deprecated `\Shopware\Storefront\Theme\ThemeLifecycleService` to be marked as final in the next major version.
```

* Changed theme configuration loading in the code, used during storefront rendering, to use the new `\Shopware\Storefront\Theme\ThemeRuntimeConfigService`:
  * `\Shopware\Storefront\Theme\ResolvedConfigLoader`
  * `\Shopware\Storefront\Theme\ThemeScripts`
  * `\Shopware\Storefront\Theme\ThemeInheritanceBuilder`
  * `\Shopware\Storefront\Framework\Routing\TemplateDataSubscriber`
* Changed `\Shopware\Storefront\Theme\CachedResolvedConfigLoaderInvalidator` name to `\Shopware\Storefront\Theme\ThemeConfigCacheInvalidator`
* Deprecated `\Shopware\Storefront\Theme\CachedResolvedConfigLoader`, as it is no longer used in the storefront
* Deprecated `\Shopware\Storefront\Theme\Exception\ThemeAssignmentException`

> * Deprecated `\Shopware\Core\System\Snippet\SnippetService::getUnusedThemes` method, replacement will not be provided.

Reviewer: maybe a small hint, how to get those data now? :thinking:

Author: Not sure that makes sense. It looks like this method was initially introduced for internal usage, new event was explicitly marked internal. But if someone needs that - they will be able to figure that out from changelog Core section.

Reviewer: ack :heavy_check_mark:

___

> # Upgrade Information

Reviewer: ```suggestion
# Upgrade Information

```

this will prevent merge conflicts :grin:

## Theme configuration changes

> * Theme configuration used during storefront rendering are now stored in a `theme_runtime_config` table and regenerated on the refresh stage of theme lifecycle.

Reviewer: ```suggestion
* Theme configuration used during storefront rendering is now stored in a `theme_runtime_config` table and regenerated on the refresh stage of theme lifecycle.
```

* The `\Shopware\Storefront\Theme\CachedResolvedConfigLoader` is now deprecated and will be removed in the next major version. Please update the code that directly uses it to use the `\Shopware\Storefront\Theme\ResolvedConfigLoader` instead.
* The `\Shopware\Storefront\Theme\Exception\ThemeAssignmentException` is now deprecated and will be removed in the next major version. Please use `\Shopware\Storefront\Theme\Exception\ThemeException::themeAssignmentException`.
___

> # Next Major Version Changes

Reviewer: we already did that on other PRs. would be nice, if you could put that in the 6.8 upgrade guide already

Author: @mitelg updated, thanks for the suggestion

## Removal of CachedResolvedConfigLoader
* The `\Shopware\Storefront\Theme\CachedResolvedConfigLoader` was removed.

## Removal of ThemeAssignmentException
* The `\Shopware\Storefront\Theme\Exception\ThemeAssignmentException` was removed.

## Changes to ThemeLifecycleService

> * The `\Shopware\Storefront\Theme\ThemeLifecycleService` became final.

Reviewer: ```suggestion
* The `\Shopware\Storefront\Theme\ThemeLifecycleService` is now final.
```
